readium/shared/src/main/java/org/readium/r2/shared/util/mediatype/MediaTypeSniffer.kt

```diff
@@ -438,44 +643,65 @@ public object EpubMediaTypeSniffer : MediaTypeSniffer {
  *  - https://www.w3.org/TR/pub-manifest/
  */
 public object LpfMediaTypeSniffer : MediaTypeSniffer {
-    override fun sniffHints(hints: MediaTypeHints): MediaType? {
+    override fun sniffHints(hints: MediaTypeHints): Try<MediaType, MediaTypeSnifferError.NotRecognized> {
         if (
             hints.hasFileExtension("lpf") ||
             hints.hasMediaType("application/lpf+zip")
         ) {
-            return MediaType.LPF
+            return Try.success(MediaType.LPF)
         }
 
-        return null
+        return Try.failure(MediaTypeSnifferError.NotRecognized)
     }
 
-    override suspend fun sniffContent(content: MediaTypeSnifferContent): MediaType? {
-        if (content !is ContainerMediaTypeSnifferContent) {
-            return null
-        }
-
-        if (content.contains("index.html")) {
-            return MediaType.LPF
+    override suspend fun sniffContainer(container: Container<Readable>): Try<MediaType, MediaTypeSnifferError> {
+        if (RelativeUrl("index.html")!! in container) {
+            return Try.success(MediaType.LPF)
         }
 
         // Somehow, [JSONObject] can't access JSON-LD keys such as `@content`.
-        content.read("publication.json")
-            ?.let { String(it) }
+        container[RelativeUrl("publication.json")!!]
+            ?.read()
+            ?.getOrElse { error ->
+                return Try.failure(MediaTypeSnifferError.Reading(error))
+            }
+            ?.let { tryOrNull { String(it) } }
             ?.let { manifest ->
                 if (
                     manifest.contains("@context") &&
                     manifest.contains("https://www.w3.org/ns/pub-context")
                 ) {
-                    return MediaType.LPF
+                    return Try.success(MediaType.LPF)
                 }
             }
 
-        return null
+        return Try.failure(MediaTypeSnifferError.NotRecognized)
+    }
+}
+
+/**
+ * Sniffs a RAR archive.
+ *
+ * At the moment, only hints are supported.
+ */
+public object RarMediaTypeSniffer : MediaTypeSniffer {
+
+    override fun sniffHints(hints: MediaTypeHints): Try<MediaType, MediaTypeSnifferError.NotRecognized> {
+        if (
+            hints.hasFileExtension("rar") ||
+            hints.hasMediaType("application/vnd.rar") ||
+            hints.hasMediaType("application/x-rar") ||
+            hints.hasMediaType("application/x-rar-compressed")
+        ) {
+            return Try.success(MediaType.LPF)
```

Comment:
A test would basically be a copy-paste from the code... I happened to see code and tests being both wrong in such cases.

Reply:
Yes, you should not copy-paste the code to write tests otherwise it won't catch any mistake. Besides that, it's useful to make sure we didn't break things when refactoring code.
Could you fill the missing tests for RAR?

Reply:
IMHO that'll be bullshit test, but okay, that's not too much work.